spring-integration-core/src/main/java/org/springframework/integration/support/MessageBuilder.java

```diff
@@ -176,7 +181,16 @@ public T getPayload() {
 	 */
 	@Override
 	public MessageBuilder<T> copyHeadersIfAbsent(Map<String, ?> headersToCopy) {
-		this.headerAccessor.copyHeadersIfAbsent(headersToCopy);
+		if (headersToCopy != null) {
+			for (Map.Entry<String, ?> entry : headersToCopy.entrySet()) {
+				if (!this.headerAccessor.isReadOnly(entry.getKey())) {
+					this.headerAccessor.setHeaderIfAbsent(entry.getKey(), entry.getValue());
+				}
+				else if (logger.isWarnEnabled()) {
+					logger.warn("The header [" + entry + "] is ignored for population because it is is readOnly.");
```

Comment:
That's true; I was thinking we'd get a lot of log noise, but that would only happen if the user keeps trying to copy that header. OK; I am happy with this, but I still think we can throw an exception while building the gateway (in 5.0).